rule/string-regex.go

```
@@ -0,0 +1,233 @@
+		return ident.Name, true
+	} else if selector, ok := call.Fun.(*ast.SelectorExpr); ok {
+		// Scoped function call
+		scope, ok := selector.X.(*ast.Ident)
+		if !ok {
+			return "", false
+		}
+		return scope.Name + "." + selector.Sel.Name, true
+	} else {
+		return "", false
+	}
+}
+
+// #endregion
+
+// #region Linting logic
+
+// Apply a single regex rule to a call expression (should be done after verifying the that the call expression matches the rule's scope)
+func (rule stringRegexSubrule) Apply(call *ast.CallExpr) {
+	if len(call.Args) <= rule.Scope.Argument {
+		// TODO: should cases where calls are incompatible with scope cause failures?
+		return
+	}
+
+	arg := call.Args[rule.Scope.Argument]
+	var lit *ast.BasicLit
+	if len(rule.Scope.Field) > 0 {
+		// Try finding the scope's Field, treating arg as a composite literal
+		composite, ok := arg.(*ast.CompositeLit)
+		if !ok {
+			return
+		}
+		for _, el := range composite.Elts {
+			kv, ok := el.(*ast.KeyValueExpr)
+			if !ok {
+				continue
+			}
+			key, ok := kv.Key.(*ast.Ident)
+			if !ok || key.Name != rule.Scope.Field {
+				continue
+			}
+
+			// We're now dealing with the exact field in the rule's scope, so if anything fails, we can safely return instead of continuing the loop
+			lit, ok = kv.Value.(*ast.BasicLit)
+			if !ok || lit.Kind != token.STRING {
+				return
+			}
+		}
+	} else {
+		var ok bool
+		// Treat arg as a string literal
+		lit, ok = arg.(*ast.BasicLit)
+		if !ok || lit.Kind != token.STRING {
+			return
+		}
+	}
+	// Unquote the string literal before linting
+	unquoted := lit.Value[1 : len(lit.Value)-1]
+	rule.lintMessage(unquoted, lit)
+}
+
+func (rule stringRegexSubrule) lintMessage(s string, node ast.Node) {
+	// Fail if the string doesn't match the user's regex
+	if rule.Regexp.MatchString(s) {
+		return
+	}
+	var failure string
+	if len(rule.ErrorMessage) > 0 {
+		failure = fmt.Sprintf("string literal doesn't match user defined regex (%s)", rule.ErrorMessage)
```

why not just report the rule error message?
i.e. in place of
`string literal doesn't match user defined regex (must start with a capital letter)`
we could just emit
`must start with a capital letter`

up to the user to provide a nice error message